### PR TITLE
Add Dockerfile and some cleanup

### LINKED ISSUE #9: Docker container to try out Coder.
I think you should add a docker container deployment one liner to your landing page similar to https://www.theia-ide.org/

Should be simple enough to create the image and add to your CI and keep an official one with Docker Hub...with docker for windows this also *sorta* meets your windows goals

### PR DESCRIPTION
**Pull request by nhooyr**
Closes #9 #42

### CODE REVIEW

File: scripts/build.sh [hunk 1]
```diff
@@ -1,5 +1,4 @@
 #!/bin/bash
 set -e
 
-npm install -g cross-env
```
Comment (kylecarbs) [inline scripts/build.sh]: Why was this removed?

File: packages/runner/src/runner.ts [hunk 2]
```diff
@@ -90,10 +93,21 @@ export const run = (name: string = process.argv[2]): void | Promise<void> => {
 			cwd = path;
 		},
 		execute(command: string, args: string[] = [], env?: object): Promise<CommandResult> {
-			return execute(command, args, {
+			const prom = execute(command, args, {
 				cwd,
 				env: env as NodeJS.ProcessEnv,
 			}, log);
+			prom.then((result: CommandResult) => {
```
Comment (kylecarbs) [inline packages/runner/src/runner.ts]: Should return this instead of prom below

File: packages/runner/src/runner.ts [hunk 3]
```diff
@@ -44,6 +46,7 @@ export type TaskFunction = (runner: Runner) => void | Promise<void>;
 
 export interface Runner {
 	cwd: string;
+
```
Comment (kylecarbs) [inline packages/runner/src/runner.ts]: Unneeded space, but not a biggie

File: scripts/build.sh (hunk 1 above)
Comment (nhooyr) [reply scripts/build.sh]: Wasn't necessary to install globally according to @coderasher as its a library.

File: packages/runner/src/runner.ts (hunk 2 above)
Comment (nhooyr) [reply packages/runner/src/runner.ts]: Interesting. It seems to work both ways. Is it a convenience thing?

File: packages/runner/src/runner.ts (hunk 3 above)
Comment (nhooyr) [reply packages/runner/src/runner.ts]: Goland formatted automatically, will fix.

File: packages/runner/src/runner.ts (hunk 2 above)
Comment (nhooyr) [reply packages/runner/src/runner.ts]: screen shot 2019-03-06 at 6 12 36 pm

Can't.

File: scripts/build.sh (hunk 1 above)
Comment (coderasher) [reply scripts/build.sh]: Yup it's part of the root `package.json`.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+RUN yarn task build:server:binary
+
+# We deploy with ubuntu so that devs have a familiar environemnt.
+FROM ubuntu
+RUN apt-get update
+RUN apt-get install -y openssl
+WORKDIR /root/project
+COPY --from=0 /src/packages/server/cli-linux /usr/local/bin/code-server
+EXPOSE 8443
+# Unfortunately `.` does not work with code-servr
```
Comment (nol166) [inline Dockerfile]: small typo

Comment (nhooyr) [reply Dockerfile]: Fixed.

File: packages/runner/src/runner.ts (hunk 2 above)
Comment (kylecarbs) [reply packages/runner/src/runner.ts]: You can add
```
return result;
```
to the end of that

Comment (nhooyr) [reply packages/runner/src/runner.ts]: Not sure what you mean. To the end of what?

Comment (kylecarbs) [reply packages/runner/src/runner.ts]: Inside the `prom.then` block

Comment (nhooyr) [reply packages/runner/src/runner.ts]: ah very fancy.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+# In the future, we can use https://github.com/yarnpkg/rfcs/pull/53 to make it use the node_modules
+# directly which should be faster.
+WORKDIR /src
+COPY . .
+RUN yarn
+RUN yarn task build:server:binary
+
+# We deploy with ubuntu so that devs have a familiar environemnt.
+FROM ubuntu
```
Comment (sr229) [inline Dockerfile]: Pin the version - since this will always resolve to `18.10` (current `latest`).

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+RUN apt-get update
+RUN apt-get install -y openssl
+WORKDIR /root/project
```
Comment (sr229) [inline Dockerfile]: I do not recommend using `root`. Consider using a actual user or a pseudo-user like what we do in Theia.

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+# In the future, we can use https://github.com/yarnpkg/rfcs/pull/53 to make it use the node_modules
+# directly which should be faster.
+WORKDIR /src
+COPY . .
+RUN yarn
+RUN yarn task build:server:binary
+
+# We deploy with ubuntu so that devs have a familiar environemnt.
+FROM ubuntu
+RUN apt-get update
```
Comment (sr229) [inline Dockerfile]: Compress these two `RUN`s as one instruction.

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,25 @@
+FROM node:8.15.0
+
+# Install VS Code's deps. These are the only two it seems we need.
+RUN apt-get update
+RUN apt-get install -y libxkbfile-dev libsecret-1-dev
```
Comment (sr229) [inline Dockerfile]: Compress these two `RUN`s as one instruction as well.

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+# Ensure latest yarn.
+RUN npm install -g yarn
+
+# In the future, we can use https://github.com/yarnpkg/rfcs/pull/53 to make it use the node_modules
+# directly which should be faster.
+WORKDIR /src
+COPY . .
+RUN yarn
```
Comment (sr229) [inline Dockerfile]: compress this as one instruction as well.

File: Dockerfile (hunk 7 above)
Comment (nhooyr) [reply Dockerfile]: Its a small thing but I like em separate to make better use of docker's layer cache.

File: Dockerfile (hunk 6 above)
Comment (nhooyr) [reply Dockerfile]: I personally don't really see much value in that. Especially for a dev setup, you pretty much need sudo 100% of the time at which point it just becomes annoying and repetitive.

cc @kylecarbs

Comment (kylecarbs) [reply Dockerfile]: I'm not certain what purpose there'd be in limiting the user when running inside the docker container. What do you think @sr229 ?

Comment (sr229) [reply Dockerfile]: @kylecarbs @nhooyr we're excercising a good practice that the user should be aware that some administrative actions might need `sudo` or a manual switch to the root. Not all Cloud IDE users are proficient in Linux and know the security implications. 

This is also how Theia and Cloud9 does their images.

File: Dockerfile (hunk 7 above)
Comment (sr229) [reply Dockerfile]: Optimizing layers no matter how small should be done.

File: Dockerfile (hunk 6 above)
Comment (nhooyr) [reply Dockerfile]: Lets get this PR in for now and discuss this in #65.

File: Dockerfile (hunk 7 above)
Comment (frol) [reply Dockerfile]: @nhooyr it is fine to separate them while you are getting the image done, but once it is there, it is [strongly recommended (see “apt-get” section)](
https://docs.docker.com/develop/develop-images/dockerfile_best-practices/) to combine update + install + remove caches

Comment (nhooyr) [reply Dockerfile]: That is interesting. Does seem reasonable to me to bust the cache every time the package list is modified.
